lewis/adapters/stream.py

```diff
@@ -163,8 +186,12 @@ def __init__(self, func, pattern, argument_mappings=None, return_mapping=None, d
             raise RuntimeError('Can not construct a Func-object from a non callable object.')
 
         self.func = func
-        self.raw_pattern = pattern
-        self.pattern = re.compile(b(pattern), 0) if pattern else None
+
+        self.raw_pattern = pattern.raw_pattern
+        self.pattern = re.compile(b(pattern.regex_pattern), 0)
```

Comment:
>Number 4 is totally unexpected and I never saw anything of that magnitude coming :)

Yes, that was a bit of a real-life plot-twist moment when I found that.

Number 3 was also a bit of a curve ball. Regarding "why we have `^` in all our regexes", I think you're right... "we've just always done it like this". But also "it always made sense like this". And, in the spirit of number 2, I'm almost tempted to say we should treat a missing `^` as an error in a command regex (even if we don't default to `fmt` in that case). Because `^S\?$` and `S\?$` are in fact exactly the same regex. So we don't miss any possible regexes by requiring `^`. But it is immediately obvious what `^S\?$` does and it can be very surprising what `S\?$` does.

If I saw interface commands defined like this (with or without explicit `regex`)...
```python
    commands = {
        Cmd('foo', regex(r'^S\?$')),
        Cmd('bar', regex(r'T\?$')),
    }
```
... I think I would be a) very surprised by the missing `^` on `T?` and b) even more surprised it actually still behaves the same as `S?`. So, to comply with the [Principle of Least Astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), I would forbid anything _not_ starting with `^` from being considered a valid regex.

Part of me is tempted to then say "well, since something that doesn't start with `^` cannot be a regex, we're free to assume it's a scanf (as long as it's not explicitly written `regex('...')`)"... but, on the other hand, I can see how this might also lead to some surprises. Especially if the user accidentally or naively misses the `^` and writes `Cmd('bar', r'T\?$')`... assuming that `T\?$` was meant to be a scanf would also be very unexpected (though arguably easier to debug, since it wouldn't match `T?` at all and you immediately know something is wrong).

For the extensibility question, do we really want to add any more formats, ever? The main reason to add `scanf` is that is matches well with the protocol files. `regex` is the general goto option for general pattern matching. I have a hard time imagining why we would want other options. Then again, I suppose the discovery of `CmdBuilder` shows that there is interest in radically different ways of specifying command format. But that is built on top of our system and doesn't involve a string where we have to decide how to interpret it.

Regarding `%([fdix]{1}`, what about `%2d`, `%*s`, `%.3f`, etc? I think it would be considerably harder to implement reliably, and lead to much more difficult rules as to "what is a regex" and "what is a scanf". The nice thing about using just `^` and nothing else is that it is a very simple to understand and remember rule and, as long as you know it, you can see at a glance whether a string will be considered regex or scanf.

Other options we haven't talked about yet:

#### 1. Factory methods
```python
    commands = {
        Cmd.regex('foo', r'^S\?$'),
        Cmd.scanf('bar', r'T=%f'),
        Cmd.builder("get_pid", arg_sep=",", ignore=r"\r\n\s").escape("QPID?").digit().build(),
    }
```
This is essentially similar to explicitly wrapping the strings, but has a few advantages:
1. `regex` and `scanf` are under `Cmd` namespace
1. No need to import formats separately
1. No need to worry about name conflicts in global space
1. Moving them outside the braces  "unclogs" the parameter values
1. Modern editors will also give you hints to available methods this way
1. They are aligned, making it easy to "batch" change them in editors that support column edits
1. The methods are free to return instances of completely different objects transparently
1. This allows radically different construction (as shown with the `CmdBuilder` style)

The only disadvantage I see is that `Var` would be locked to the same format for both read and write. This is not necessarily a bad thing to be honest.

#### 2. Different Classes (or Factory Functions)
```python
    commands = {
        CmdRegex('foo', r'^S\?$'),
        CmdScanf('bar', r'T=%f'),
        CmdBuilder("get_pid", arg_sep=",", ignore=r"\r\n\s").escape("QPID?").digit().build(),
    }
```
Very similar to factory methods. These could be actual classes or merely capitalized functions that construct `Cmd` with different parameters. `Cmd` could still exist and default to `CmdRegex` behavior. You could "change default behavior" like this:
```python
from lewis.adapters.stream import CmdScanf as Cmd
# ...
    commands = {
        Cmd('foo', r'^S\?$'),
    }
```
The disadvantage would be losing containment under `Cmd` namespace and the advantages that come with that (code organization, encapsulation, intellisense hints, importing separately).

Also, while `CmdRegex` and `CmdScanf` have compatible signatures, `CmdBuilder` (or anything else we might want to add) works completely differently. So the alias import trick is limited. The factory method syntax being a bit stricter in that regard could be seen as an advantage.

Sorry about the huge reply... it's surprising how much can be said about such a simple feature.